### PR TITLE
Solved customfield default value not showing up upon device creation

### PR DESCRIPTION
**Pull request by HanlinMiao**
The steps to reproduce the error now creates a device with default_values populated on the custom_fields. However, there are still some bugs to work out.
Screen Shot 2022-06-13 at 10 26 30 PM
5-8d41dcd7-05f0-4a48-96cb-adb4cefd9238.png">

### CODE REVIEW

File: nautobot/dcim/models/device_component_templates.py [hunk 1]
```diff
@@ -84,7 +85,20 @@ def instantiate_model(self, model, device, **kwargs):
         """
         Helper method to self.instantiate().
         """
-        return model(device=device, name=self.name, label=self.label, description=self.description, **kwargs)
+        content_type = ContentType.objects.get_for_model(model)
+        fields = CustomField.objects.filter(content_types=content_type)
+        for field in fields:
+            self._custom_field_data.setdefault(field.name, field.default)
```
Comment (glennmatthews) [inline nautobot/dcim/models/device_component_templates.py:91]: I'm not sure that storing the instance's custom field defaults onto `self` makes sense here. In particular a ComponentTemplateModel may theoretically have its own custom fields that are not actually applicable to the related ComponentModel - probably better to just construct a dict specifically for the `model` that can be passed through to the constructor and then discarded.

File: nautobot/dcim/models/device_component_templates.py [hunk 2]
```diff
@@ -249,11 +250,35 @@ class Meta:
         unique_together = ("device_type", "name")
 
     def instantiate(self, device):
+        record = {}
+        content_type = ContentType.objects.get_for_model(Interface)
+        fields = CustomField.objects.filter(content_types=content_type)
+        for field in fields:
+            record[field.name] = field.default
+
```
Comment (HanlinMiao) [inline nautobot/dcim/models/device_component_templates.py:258]: Hi Glenn @glennmatthews, thank you for the advice. Did I understand you correctly?

Comment (glennmatthews) [reply nautobot/dcim/models/device_component_templates.py:258]: I thought your previous approach of loading the custom field data during `instantiate_model` made sense, I'm not sure I see the benefit of doing this in `instantiate` as you are now.

My point was just that the `_custom_field_data` that you pass through to the `model()` call (e.g., the `_custom_field_data` for an `Interface` instance being created) is *not* the same as `self._custom_field_data` (e.g. the `_custom_field_data` on the `InterfaceTemplate` that's creating the `Interface`). In other words, `instantiate()`/`instantiate_model()` on the `ComponentTemplateModel` should *not* modify `self._custom_field_data`, nor should `self._custom_field_data` be passed to the `model()` call.

File: nautobot/dcim/models/device_component_templates.py [hunk 3]
```diff
@@ -84,7 +85,20 @@ def instantiate_model(self, model, device, **kwargs):
...
+        return model(
+            device=device,
+            name=self.name,
+            label=self.label,
+            description=self.description,
+            _custom_field_data=record,
```
Comment (glennmatthews) [inline nautobot/dcim/models/device_component_templates.py:99]:
```suggestion
        custom_field_data = {}
        content_type = ContentType.objects.get_for_model(model)
        fields = CustomField.objects.filter(content_types=content_type)
        for field in fields:
            custom_field_data[field.name] = field.default

        return model(
            device=device,
            name=self.name,
            label=self.label,
            description=self.description,
            _custom_field_data=custom_field_data,
```

File: nautobot/dcim/models/device_component_templates.py [hunk 4]
```diff
@@ -249,12 +263,7 @@ class Meta:
         unique_together = ("device_type", "name")
 
     def instantiate(self, device):
-        return self.instantiate_model(
-            model=Interface,
-            device=device,
-            type=self.type,
-            mgmt_only=self.mgmt_only,
-        )
+        return self.instantiate_model(model=Interface, device=device, type=self.type, mgmt_only=self.mgmt_only)
```
Comment (glennmatthews) [inline nautobot/dcim/models/device_component_templates.py:266]: Revert this?

**Review by glennmatthews — COMMENTED** [review]
Code looks good. Can we get a unit test case written for this?

File: nautobot/dcim/models/device_component_templates.py [hunk 5]
```diff
@@ -14,7 +15,7 @@
 
 from nautobot.core.models import BaseModel
 from nautobot.dcim.constants import REARPORT_POSITIONS_MAX, REARPORT_POSITIONS_MIN
-from nautobot.extras.models import CustomFieldModel, ObjectChange, RelationshipModel
+from nautobot.extras.models import CustomFieldModel, ObjectChange, RelationshipModel, CustomField
```
Comment (timizuoebideri1) [inline nautobot/dcim/models/device_component_templates.py:18]:
```suggestion
from nautobot.extras.models import CustomField, CustomFieldModel, ObjectChange, RelationshipModel
```

File: nautobot/dcim/tests/test_models.py [hunk 6]
```diff
@@ -31,7 +37,62 @@
...
+        )
+        custom_field_1.content_types.set([ContentType.objects.get_for_model(Interface)])
+        custom_field_2 = CustomField.objects.create(
+            type=CustomFieldTypeChoices.TYPE_TEXT, name="field_2", default="value_2"
+        )
+        custom_field_2.content_types.set([ContentType.objects.get_for_model(Interface)])
+        custom_field_3 = CustomField.objects.create(
+            type=CustomFieldTypeChoices.TYPE_TEXT, name="field_3", default="value_3"
+        )
+        custom_field_3.content_types.set([ContentType.objects.get_for_model(Interface)])
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:65]: Suggested rework to slightly reduce duplicate code:

```suggestion
        custom_fields = [
            CustomField.objects.create(
                type=CustomFieldTypeChoices.TYPE_TEXT, name="field_1", default="value_1"
            ),
            CustomField.objects.create(
                type=CustomFieldTypeChoices.TYPE_TEXT, name="field_2", default="value_2"
            ),
            CustomField.objects.create(
                type=CustomFieldTypeChoices.TYPE_TEXT, name="field_3", default="value_3"
            ),
        ]
        for custom_field in custom_fields:
            custom_field.content_types.set([ContentType.objects.get_for_model(Interface)])
```

Also possibly consider using different `type` values for the different fields, just to make the test a bit broader (though it shouldn't matter in this particular case)?

File: nautobot/dcim/tests/test_models.py [hunk 7]
```diff
@@ -31,7 +37,62 @@
     Site,
 )
 from nautobot.extras.models import Status
+from nautobot.extras.models.customfields import CustomField
 from nautobot.tenancy.models import Tenant
+from nautobot.extras.choices import CustomFieldTypeChoices
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:42]:
```suggestion
from nautobot.extras.choices import CustomFieldTypeChoices
from nautobot.extras.models import CustomField, Status
from nautobot.tenancy.models import Tenant
```

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**